### PR TITLE
storage, txn_types: output previous write record to engine during prewrite

### PR DESCRIPTION
**Pull request by 5kbpers**
### What problem does this PR solve?

Problem Summary:
Old Value is the previously written value before a delete/update operation.
To support the new collation format and compatible with TiDB Binlog, TiCDC requires TiKV to output the old value for it.

### What is changed and how it works?

Separate from #8071.

What's Changed:
* Add `extra_op` field to `MvccTxn`
* check and output the previous write record to `Engine` as `TxnExtra` during prewrite

### Related changes

- Should be cherry-picked to release 4.0 after all related PRs got merged

### Check List 

Tests 

- Unit test
- Integration test

Side effects

- Performance regression
    - Consumes more CPU
    - Consumes more MEM

### Release note 
* N/A

### CODE REVIEW

File: components/txn_types/src/types.rs [hunk 1]
```diff
@@ -292,6 +312,46 @@ impl From<kvrpcpb::Mutation> for Mutation {
     }
 }
 
+// Returned by MvccTxn when extra_op is set to kvrpcpb::ExtraOp::ReadOldValue.
+// key with current ts -> (short value of the prev txn, start ts of the prev txn).
+// the value is None when the mutation is `Insert`.
```
Comment (nrc) [inline components/txn_types/src/types.rs:317]: Could you clarify the comment, there are two Options which could be None. (Even better if we can get rid of one :-) )

File: components/txn_types/src/types.rs [hunk 2]
```diff
@@ -292,6 +312,46 @@ impl From<kvrpcpb::Mutation> for Mutation {
...
+pub type OldValues = HashMap<Key, Option<(Option<Value>, TimeStamp)>>;
+
+// Extra data fields filled by kvrpcpb::ExtraOp.
+#[derive(Default, Debug, Clone)]
+pub struct TxnExtra {
+    old_values: Option<OldValues>,
```
Comment (nrc) [inline components/txn_types/src/types.rs:323]: Do we need an `Optinon` here? I think we could just use an empty `HashMap`.

File: components/txn_types/src/types.rs [hunk 3]
```diff
@@ -292,6 +312,46 @@ impl From<kvrpcpb::Mutation> for Mutation {
...
+            }
+        };
+        old_values.insert(key, value);
+    }
+
+    pub fn is_empty(&mut self) -> bool {
+        self.old_values.is_none()
+    }
+
+    pub fn append(&mut self, other: &mut Self) {
```
Comment (nrc) [inline components/txn_types/src/types.rs:342]:
```suggestion
    pub fn extend(&mut self, other: &mut Self) {
```

File: components/txn_types/src/types.rs [hunk 4]
```diff
@@ -292,6 +312,46 @@ impl From<kvrpcpb::Mutation> for Mutation {
...
+        if let Some(old_values) = other.old_values.as_mut() {
+            for (k, v) in old_values.drain() {
```
Comment (nrc) [inline components/txn_types/src/types.rs:344]: I think you can use `HashMap::extend` rather than a for loop here.

File: src/server/raftkv.rs [hunk 5]
```diff
@@ -148,6 +148,7 @@ fn on_read_result(
     mut read_resp: ReadResponse<RocksSnapshot>,
     req_cnt: usize,
 ) -> (CbContext, Result<CmdRes>) {
+    // TODO(5kbpers): set ExtraOp for cb_ctx here.
```
Comment (nrc) [inline src/server/raftkv.rs:151]: Should this be fixed before landing?

File: src/storage/mvcc/reader/scanner/mod.rs [hunk 6]
```diff
@@ -359,6 +359,41 @@ pub fn has_data_in_range<S: Snapshot>(
     Ok(false)
 }
 
+pub fn seek_for_valid_write<I>(
```
Comment (nrc) [inline src/storage/mvcc/reader/scanner/mod.rs:362]: Please add a comment

File: src/storage/mvcc/txn.rs [hunk 7]
```diff
@@ -96,7 +99,15 @@ impl<S: Snapshot> MvccTxn<S> {
     }
 
     pub fn into_modifies(self) -> Vec<Modify> {
-        self.writes
+        self.writes.modifies
+    }
+
+    pub fn set_extra_op(&mut self, extra_op: ExtraOp) {
```
Comment (nrc) [inline src/storage/mvcc/txn.rs:105]: Make `extra_op` public so we don't need this method.

File: src/storage/mvcc/txn.rs [hunk 8]
```diff
@@ -1003,6 +1018,44 @@ impl<S: Snapshot> MvccTxn<S> {
             is_completed,
         })
     }
+
+    fn check_extra_op(
```
Comment (nrc) [inline src/storage/mvcc/txn.rs:1022]: Please add a comment

File: src/server/raftkv.rs (hunk 5 above)
Comment (5kbpers) [reply src/server/raftkv.rs:151]: It will be addressed at subsequent PRs.

**Comment by 5kbpers** [comment]
PTAL, thanks!

File: src/storage/txn/process.rs [hunk 9]
```diff
@@ -593,12 +609,13 @@ fn process_write_impl<S: Snapshot, L: LockManager>(
             statistics.add(&txn.take_statistics());
             if locks.is_empty() {
                 let pr = ProcessResult::MultiRes { results: vec![] };
-                let modifies = txn.into_modifies();
-                (pr, modifies, rows, cmd.ctx, None)
+                let txn_extra = txn.take_extra();
+                let write_data = WriteData::new(txn.into_modifies(), txn_extra);
```
[the remark this answers is not in the document]
Comment (5kbpers) [reply src/storage/txn/process.rs:613]: Pessimistic prewrite doesn't read the previous write record for checking conflict, so we must read the old value in other places (like cdc threads). But I just realized we can put the `mutation_type` to the map of old values, then some unnecessary reads could be avoided.

File: src/storage/txn/process.rs [hunk 10]
```diff
@@ -636,12 +653,12 @@ fn process_write_impl<S: Snapshot, L: LockManager>(
             statistics.add(&txn.take_statistics());
             if locks.is_empty() {
                 let pr = ProcessResult::MultiRes { results: vec![] };
-                let modifies = txn.into_modifies();
-                (pr, modifies, rows, cmd.ctx, None)
+                let write_data = WriteData::from_modifies(txn.into_modifies());
```
Comment (MyonKeminta) [inline src/storage/txn/process.rs:656]: Is extra op set for `txn` in pessimistic prewrite?

**Pull request merged**